ml-agents/mlagents/trainers/ppo/models.py

```diff
@@ -79,195 +77,77 @@ def create_reward_encoder():
-        else:
-            pred_action = tf.concat(
-                [
-                    tf.layers.dense(hidden, self.act_size[i], activation=tf.nn.softmax)
-                    for i in range(len(self.act_size))
-                ],
-                axis=1,
-            )
-            cross_entropy = tf.reduce_sum(
-                -tf.log(pred_action + 1e-10) * self.selected_actions, axis=1
-            )
-            self.inverse_loss = tf.reduce_mean(
-                tf.dynamic_partition(cross_entropy, self.mask, 2)[1]
-            )
-
-    def create_forward_model(self, encoded_state, encoded_next_state):
-        """
-        Creates forward model TensorFlow ops for Curiosity module.
-        Predicts encoded future state based on encoded current state and given action.
-        :param encoded_state: Tensor corresponding to encoded current state.
-        :param encoded_next_state: Tensor corresponding to encoded next state.
-        """
-        combined_input = tf.concat([encoded_state, self.selected_actions], axis=1)
-        hidden = tf.layers.dense(combined_input, 256, activation=self.swish)
-        # We compare against the concatenation of all observation streams, hence `self.vis_obs_size + int(self.vec_obs_size > 0)`.
-        pred_next_state = tf.layers.dense(
-            hidden,
-            self.curiosity_enc_size * (self.vis_obs_size + int(self.vec_obs_size > 0)),
-            activation=None,
-        )
-
-        squared_difference = 0.5 * tf.reduce_sum(
-            tf.squared_difference(pred_next_state, encoded_next_state), axis=1
-        )
-        self.intrinsic_reward = tf.clip_by_value(
-            self.curiosity_strength * squared_difference, 0, 1
-        )
-        self.forward_loss = tf.reduce_mean(
-            tf.dynamic_partition(squared_difference, self.mask, 2)[1]
-        )
-
-    def create_ppo_optimizer(
-        self, probs, old_probs, value, entropy, beta, epsilon, lr, max_step
+    def create_losses(
+        self,
+        probs,
+        old_probs,
+        value_streams,
+        entropy,
+        beta,
+        epsilon,
+        lr,
+        max_step,
+        stream_names,
     ):
         """
         Creates training-specific Tensorflow ops for PPO models.
         :param probs: Current policy probabilities
         :param old_probs: Past policy probabilities
-        :param value: Current value estimate
+        :param value_streams: Current value estimates from each value stream
         :param beta: Entropy regularization strength
         :param entropy: Current policy entropy
         :param epsilon: Value for policy-divergence threshold
         :param lr: Learning rate
         :param max_step: Total number of training steps.
         """
-        self.returns_holder = tf.placeholder(
-            shape=[None], dtype=tf.float32, name="discounted_rewards"
-        )
+        self.returns_holders = []
+        self.old_values = []
+        for i in range(len(value_streams)):
```

Comment:
I'm a little confused on the types/uses of `value_streams` and `stream_names`. Former is a dict of str->some tf object, and latter is a list of str? Is `stream_names` a subset of `value_streams.keys()`? Can you have a `stream_name` that's not a key in `value_streams`?

If `stream_names` is a subset of `value_streams.keys()` then it feels strange to iterate over the `len(value_streams)` here. 

If you can have a stream_name that's not a key in value_streams, the `for i, name in enumerate(stream_names):` seems wrong because it uses them as keys.

If they're identical, then you should drop stream_names as a parameter and just `value_streams.keys()`

Reply:
I re-did this - now I'm always referring to things by stream name and never by index. Also removed stream_names and just use the keys from the `value_streams` (now `value_heads`) parameter. 

The `value_streams` aren't the same as the `stream_names` - renamed `value_streams` to `value_heads` (they're Tensorflow nodes) and removed stream_names. Let me know how that looks!

Reply:
Sounds cleaner, will look again tomorrow.